src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1844,4 +1845,100 @@ impl Cage {
 
         0 //getcwd has succeeded!;
     }
+
+    //------------------SHMGET SYSCALL------------------
+
+    pub fn shmget_syscall(&self, key: i32, size: usize, shmflg: i32)-> i32 {
+        if key == IPC_PRIVATE {return syscall_error(Errno::ENOENT, "shmget", "IPC_PRIVATE not implemented");}
+        if (size as u32) < SHMMIN || (size as u32) > SHMMAX { return syscall_error(Errno::EINVAL, "shmget", "Size is less than SHMMIN or more than SHMMAX"); }
+        let shmid: i32;
+        let metadata = &SHM_METADATA;
+
+        match metadata.shmkeyidtable.entry(key) {
+            interface::RustHashEntry::Occupied(occupied) => {
+                if (IPC_CREAT | IPC_EXCL) == (shmflg & (IPC_CREAT | IPC_EXCL)) {
+                    return syscall_error(Errno::EEXIST, "shmget", "key already exists and IPC_CREAT and IPC_EXCL were used");
+                }
+                shmid = *occupied.get(); 
+            }
+            interface::RustHashEntry::Vacant(vacant) => {
+                if 0 == (shmflg & IPC_CREAT) {
+                    return syscall_error(Errno::ENOENT, "shmget", "tried to use a key that did not exist, and IPC_CREAT was not specified");
+                }
+                drop(vacant);
+                shmid = metadata.new_keyid(key);
+                let mode = (shmflg & 0x1FF) as u16; // mode is 9 least signficant bits of shmflag, even if we dont really do anything with them
+
+                let segment = new_shm_segment(key, size, self.cageid as u32, DEFAULT_UID, DEFAULT_GID, mode);
+                metadata.shmtable.insert(shmid, segment);
```

Comment:
This is not the correct way to use the entry API here.
`new_keyid` should give you a new id but not actually insert it into the table.
You should use `vacant.insert(shmid)` to actually populate the entry

Reply:
Ah ok.